WordPress/openverse-frontend
# Add support for native TypeScript

## Fixes


Fixes #921 by @sarayourfriend

## Description


Adds support for native TypeScript by adding the following dependencies:

* `@nuxt/typescript-build` -> This adds TypeScript to our build pipeline
* `@typescript-eslint/*` -> These packages enable TypeScript parsing and recommended rules for ESLint
* `@babel/preset-typescript` -> This adds TypeScript to the babel configuration we use for running `jest`

Additionally, I've had to add a `.pnpmfile.cjs` ([documentation for that here](https://pnpm.io/pnpmfile)) in order to force resolve `typescript` to a modern version. We need this because `@nuxt/typescript-build` uses an older version of TypeScript that doesn't support some of the TS features we were already using (like casting to `const` with JSDoc and type parameter defaults in JSDoc).

Finally, I've converted `deepFreeze` to be native TypeScript to prove that this works top to bottom.

## Testing Instructions

Checkout the branch and run `pnpm dev`. Ensure the application compiles with no errors. Visit the search results page to make sure that `deepFreeze` is actually being run and verify that there are no errors.

Make some changes to the code that will fail TypeScript compilation. For example, in `deepFreeze.ts` add the following:

```ts
const n: number = '123'
```

Confirm that `pnpm types` fails and that `pnpm dev` spits out a compilation error.

Finally, test the ESLint integration by adding something that will fail eslint in a TypeScript file and verify that the error shows.

The jest integration is tested by the tests passing locally and in CI. Ensure they pass locally by running `pnpm test` and check that the CI passes them as well.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
